### ISSUE TITLE
cant build memcached docker image for alpine, unit test falls

### ISSUE THREAD

**Issue by cyrax13** [memcached/memcached#861]
Hi all)

I am using official dockerfile to build memcached in docker. My way:
Link to dockerhub: https://hub.docker.com/_/memcached/
Then i get link to dockerfile from section _Supported tags and respective Dockerfile links_ based on alpine linux and download: https://github.com/docker-library/memcached/blob/899756d891c484fba28c1e29b488e407c0e2d61b/alpine/Dockerfile

**Reproduce**
Download files: Dockerfile + docker-entrypoint.sh from links above
Run command to build docker image:
`docker build -f Dockerfile -t memcached:1.6.13 .`

**Results**
Memcached assembles successfully. After build the unit tests runs and they fall on the test _watcher.t_.
Details from log:
```
t/touch.t ................... ok
t/unixsocket.t .............. ok
t/whitespace.t .............. skipped: Skipping test because this does not appear to be a memcached git working directory
t/watcher_connid.t .......... ok
#   Failed test 'logged closed connection due to client disconnect'
#   at t/watcher.t line 95.
#                   'ts=1643017719.325765 gid=3 type=conn_new rip=unix rport=0 transport=local cfd=25
# 
#     doesn't match '(?^:ts=\d+\.\d+\ gid=\d+ type=conn_close .+ transport=(local|tcp) reason=normal)'
#   Failed test 'logged new connection'
#   at t/watcher.t line 102.
#                   'ts=1643017719.326001 gid=4 type=conn_close rip=unix rport=0 transport=local reason=normal cfd=25
# 
#     doesn't match '(?^:ts=\d+\.\d+\ gid=\d+ type=conn_new .+ transport=(local|tcp))'
#   Failed test 'logged closed connection due to client protocol error'
#   at t/watcher.t line 105.
#                   'ts=1643017719.330355 gid=5 type=conn_new rip=unix rport=0 transport=local cfd=25
# 
#     doesn't match '(?^:ts=\d+\.\d+\ gid=\d+ type=conn_close .+ transport=(local|tcp) reason=error)'
# Looks like you failed 3 tests of 30.
t/watcher.t ................. 
Dubious, test returned 3 (wstat 768, 0x300)
Failed 3/30 subtests
```
Full output log here: [memcached_build_log.txt](https://github.com/memcached/memcached/files/7951374/memcached_build_log.txt)

Due to this error the docker image is not built.

The bug appeared in version 1.6.12 and 1.6.13. Not noticed in versions 1.6.10 and below.
Possibly after editing the file with the unit test according to the revision https://github.com/memcached/memcached/commit/32eab461f58ce5949baa7ffd47b0c6cefbbb1bd4
Can this issue be fixed?

**System Information**
 - OS/Distro: Alpine Linux
 - Version of OS/distro: 3.15
 - Version of memcached: 1.6.13 current (the same failed build was at 1.6.12)
 - Hardware detail: virtual machine with Ubuntu Server LTS 20.04 + Docker 20.10 last version, run on VMWare at host with Windows 10x64, i7-8700k, 32gb ddr4...

**Comment by dormando** [comment]
Thanks! I'll look into this soon. @LINKIWI - if you want, as well.

I've also been getting a watcher.t failure in my centos 7 builder that's persistent but not obvious. might've introduced a few bugs into the watch code lately.

**Comment by LINKIWI** [comment]
I just tried and was able to build that Docker image on a few different servers. That said though I remember running into some race condition issues in `watcher.t` a ~few months ago though I haven't been able to reproduce them since. If I have some time I'll try to investigate as well.

**Comment by dormando** [comment]
Thanks! Fwiw the centos7 error:
```#   Failed test 'found rawcmd log entry'
#   at t/watcher.t line 140.
#          got: '0'
#     expected: '1'
#   Failed test 'found eviction log entry'
#   at t/watcher.t line 141.
#          got: '0'
#     expected: '1'
Can't use an undefined value as a symbol reference at t/watcher.t line 175.
# Looks like you planned 30 tests but ran 16.
# Looks like you failed 2 tests of 16 run.
# Looks like your test exited with 111 just after 16.
t/watcher.t .................
```

Also got on freebsd-12:
```
#   Failed test 'logged closed connection due to client protocol error'
#   at t/watcher.t line 105.
#                   'ts=1643266200.322632 gid=5 type=conn_new rip=unix rport=0 transport=local cfd=19
# '
#     doesn't match '(?^:ts=\d+\.\d+\ gid=\d+ type=conn_close .+ transport=(local|tcp) reason=error)'
# Looks like you failed 1 test of 30,
```

Looks like I haven't updated any/most of the buildbots to also run the TLS tests... so that might not be seeing this potentially TLS-specific issue with docker.

**Comment by dormando** [comment]
Doing another release today-ish. Should probably try to look at this..

**Comment by dormando** [comment]
If I run the test in a loop long enough on my machine it does fail:
```
ok 7 - connevents watcher enabled
RES1: ts=1644442913.429289 gid=2 type=conn_new rip=unix rport=0 transport=local cfd=23

ok 8 - logged new connection
RES2: ts=1644442913.429725 gid=3 type=conn_new rip=unix rport=0 transport=local cfd=25

not ok 9 - logged closed connection due to client disconnect
#   Failed test 'logged closed connection due to client disconnect'
#   at t/watcher.t line 97.
#                   'ts=1644442913.429725 gid=3 type=conn_new rip=unix rport=0 transport=local cfd=25
# '
#     doesn't match '(?^:ts=\d+\.\d+\ gid=\d+ type=conn_close .+ transport=(local|tcp) reason=normal)'
RES3: ts=1644442913.429763 gid=4 type=conn_close rip=unix rport=0 transport=local reason=normal cfd=25

not ok 10 - logged new connection
#   Failed test 'logged new connection'
#   at t/watcher.t line 105.
#                   'ts=1644442913.429763 gid=4 type=conn_close rip=unix rport=0 transport=local reason=normal cfd=25
# '
#     doesn't match '(?^:ts=\d+\.\d+\ gid=\d+ type=conn_new .+ transport=(local|tcp))'
RES4: ts=1644442913.430898 gid=5 type=conn_new rip=unix rport=0 transport=local cfd=25

not ok 11 - logged closed connection due to client protocol error
#   Failed test 'logged closed connection due to client protocol error'
#   at t/watcher.t line 109.
#                   'ts=1644442913.430898 gid=5 type=conn_new rip=unix rport=0 transport=local cfd=25
# '
#     doesn't match '(?^:ts=\d+\.\d+\ gid=\d+ type=conn_close .+ transport=(local|tcp) reason=error)'
Signal handled: Interrupt.
ok 12 - new watcher enabled
```

Added prints to the result it's checking. I see two different fds for new conns. Looks like what it's potentially seeing are the log entries for the watcher socket itself, or the client socket that gets created by the `new_memcached()` code.

The logger's async nature makes it pretty awkward to test and there're no commands to force a flush or synchronization.

I wonder what a more reliable method for testing the watcher would be?

I added a `sleep 1;` between creating the watcher socket and running the watcher command and am running the test loop again. This might be the same fix for the other errors I was seeing, but it's still a race until we get a way of testing syncs. Maybe a global counter for how often the logs are checked, and test clients wait for that to tick up a couple times before starting a watch command? @LINKIWI

**Comment by LINKIWI** [comment]
Yeah the async logging makes it pretty hairy to test. Seems like there are at least a few other occurrences of `sleep n;` in the tests to work around similar race conditions so I suppose it would suffice for band-aid solution.

Looks like there's a [`log_watcher_sent` counter](https://github.com/memcached/memcached/blob/master/logger.c#L713) that unit test clients could poll on? (If I'm understanding the proposal correctly)

**Comment by dormando** [comment]
The `sleep 1;` change ran for an hour without hanging (though it's not hanging at the same place my centos failure was), so I'll go with that for now.

Just checked `log_watcher_sent` - that counts how many entries were sent to actual watchers.. so if you're trying to poll for when to start a watcher on a clean slate I'm not sure that's it.

Looks like I may have made this worse with d63db64889952b160036e90bdb1eba024cf33712 - since the logger won't advance without active watchers. This change does try to ensure that old entries don't get written out to a new watcher, so the problem we see in tests is still the async logs getting picked up at the wrong time.

Could just skip the cond wait `#ifndef NDEBUG`  - maybe also crank down `MAX_LOGGER_SLEEP` in debug to shorten the annoyance window.

**Comment by dormando** [comment]
Alright, I pushed a workaround. The centos7 failure is actually a segfault, but in my infinite wisdom I lost the root passwords to the VM hosts (despite copious notes for the buildbot system...) and it's segfaulting within glibc so I have no idea what's going on there beyond that it started around late september.

Will close out this issue as this is fixed for the next release now. Will do separate issue/PR if me (or anyone else) has a better idea of how to structure these tests.
